### PR TITLE
Custom validators module

### PR DESCRIPTION
**Pull request by vgorkavenko**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,6 +1,6 @@
 # ethereum-validators-monitoring (aka balval)
 
-Consensus layer validators monitoring bot, that fetches Lido Node Operator keys
+Consensus layer validators monitoring bot, that fetches Lido or User Node Operator keys
```
Comment (dgusakov) [inline README.md:3]:
```suggestion
Consensus layer validators monitoring bot, that fetches Lido or Custom Users Node Operators keys
```

Comment (dgusakov) [reply README.md:3]: Also add a short note on how to implement custom registry module

Comment (vgorkavenko) [reply README.md:3]: > Also add a short note on how to implement custom registry module

https://github.com/lidofinance/ethereum-validators-monitoring/blob/feat/custom-validators/README.md#use-custom-validators-list

**Review by dgusakov — APPROVED** [review]

File: src/common/alertmanager/critical-alerts.module.ts [hunk 2]
```diff
@@ -1,10 +1,10 @@
 import { Module } from '@nestjs/common';
 import { CriticalAlertsService } from './critical-alerts.service';
-import { StorageModule } from '../../storage';
 import { PrometheusModule } from '../prometheus';
+import { ClickhouseModule } from '../../storage/clickhouse';
```
Comment (madlabman) [inline src/common/alertmanager/critical-alerts.module.ts:4]:
```suggestion
import { ClickhouseModule } from 'storage/clickhouse';
```

and other ones

**Review by madlabman — CHANGES_REQUESTED** [review]

File: README.md [hunk 3]
```diff
@@ -27,50 +27,62 @@ chown -R 472:472 .volumes/grafana
...
+6. Run `yarn start:prod`
+
+## Use custom validators list
+
+By default, monitoring bot fetches validator keys from Lido contract, but you can monitor your own validators:
+1. Set `file` value to `VALIDATOR_REGISTRY_SOURCE` env var
+2. Create file with keys by example [here](docker/validators/custom_mainnet.yaml)
+3. Set `<path to your file>` to `VALIDATOR_REGISTRY_FILE_SOURCE_PATH` env var
+
+If you want to implement your own source, it must match [RegistrySource interface](src/common/validators-registry/registry-source.interface.ts) and be included in [RegistryModule providers](src/common/validators-registry/registry.module.ts)
```
[the remark this answers is not in the document]
Comment (vgorkavenko) [reply README.md:47]: It can be a simple parser from off-chain url, for example

[the remark this answers is not in the document]
Comment (madlabman) [reply README.md:47]: Got you

File: src/common/alertmanager/critical-alerts.module.ts (hunk 2 above)
Comment (vgorkavenko) [reply src/common/alertmanager/critical-alerts.module.ts]: fixed

File: src/inspector/processing/stats-processing.service.ts [hunk 4]
```diff
@@ -1,18 +1,18 @@
...
-import { ConfigService } from '../../common/config';
-import { Owner, PrometheusService, PrometheusValStatus } from '../../common/prometheus';
-import { RegistryService } from '../../common/validators-registry';
-import { LIDO_CONTRACT_TOKEN, Lido } from '@lido-nestjs/contracts';
-import { ValidatorsStatusStats } from '../../storage/clickhouse';
+import { ConfigService } from 'common/config';
+import { Owner, PrometheusService, PrometheusValStatus } from 'common/prometheus';
+import { RegistryService } from 'common/validators-registry';
+import { ValidatorsStatusStats } from 'storage/clickhouse';
+import { LidoSourceService } from '../../common/validators-registry/lido-source';
```
[the remark this answers is not in the document]
Comment (vgorkavenko) [reply src/inspector/processing/stats-processing.service.ts]: fixed

File: src/common/validators-registry/file-source/file-source.service.ts [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+@Injectable()
+export class FileSourceService implements RegistrySource {
+  constructor(@Inject(LOGGER_PROVIDER) protected readonly logger: LoggerService, protected readonly configService: ConfigService) {}
+
+  protected data: FileContent;
+  protected lastSuccessDataReadTimestamp: number;
+
+  public async update() {
+    const fileContent = readFileSync(this.configService.get('VALIDATOR_REGISTRY_FILE_SOURCE_PATH'), 'utf-8');
```
[the remark this answers is not in the document]
Comment (vgorkavenko) [reply src/common/validators-registry/file-source/file-source.service.ts]: I don't know. Changed to the correct function

File: src/common/validators-registry/file-source/file-source.service.ts [hunk 6]
```diff
@@ -0,0 +1,84 @@
+import { Inject, Injectable, LoggerService } from '@nestjs/common';
+import {
+  RegistrySource,
+  RegistrySourceKey,
+  RegistrySourceKeysIndexed,
+  RegistrySourceKeyWithOperatorName,
```
[the remark this answers is not in the document]
Comment (vgorkavenko) [reply src/common/validators-registry/file-source/file-source.service.ts]: fixed. now just it uses `RegistrySourceKeyWithOperatorName` in file source module

File: src/inspector/processing/stats-processing.service.ts [hunk 7]
```diff
@@ -168,8 +153,25 @@ export class StatsProcessingService implements OnModuleInit {
...
+        operators.reduce((sum, o: RegistryOperator) => sum + o.usedSigningKeys, 0),
+      );
+      // only for operators with 0 used keys
+      operators.forEach((operator: RegistryOperator) => {
+        if (operator.usedSigningKeys == 0) {
+          this.prometheus.userValidators.set({ nos_name: operator.name, status: PrometheusValStatus.Ongoing }, 0);
+        }
+      });
+
+      const bufferedEther = (await this.registryService.source.contract.getBufferedEther()).div(1e9).div(1e9);
```
[the remark this answers is not in the document]
Comment (vgorkavenko) [reply src/inspector/processing/stats-processing.service.ts]: named

**Review by madlabman — APPROVED** [review]

**Pull request merged**